test/extensions/transport_sockets/internal_upstream/internal_upstream_integration_test.cc

```diff
@@ -182,5 +191,399 @@ TEST_F(InternalUpstreamIntegrationTest, BasicFlowWithoutTransportSocket) {
+    auto* filter = filter_chain->add_filters();
+    filter->set_name("envoy.filters.network.http_connection_manager");
+    filter->mutable_typed_config()->PackFrom(hcm);
+  });
+}
+
+// Setup the external upstream which has the DFP cluster.
+void HttpsInspectionIntegrationTest::setupExternalUpstreamConfig() {
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* static_resources = bootstrap.mutable_static_resources();
+    auto* cluster_0 = static_resources->mutable_clusters(0);
+    cluster_0->mutable_cluster_type()->set_name("envoy.clusters.dynamic_forward_proxy");
+    cluster_0->set_lb_policy(envoy::config::cluster::v3::Cluster_LbPolicy_CLUSTER_PROVIDED);
+
+    // Setup DFP ClusterConfig.
+    envoy::extensions::clusters::dynamic_forward_proxy::v3::ClusterConfig dfp_cluster_config;
+    auto *dns_cache_config = dfp_cluster_config.mutable_dns_cache_config();
+    setupDnsCacheConfig(*dns_cache_config);
+    cluster_0->mutable_cluster_type()->mutable_typed_config()->PackFrom(dfp_cluster_config);
+    cluster_0->clear_load_assignment();
+  });
+}
+
+void HttpsInspectionIntegrationTest::setupTestConfig() {
+  setupBootstrapExtension();
+  setupExternalListenerConfig();
+  setupInternalUpstreamConfig();
+  setupInternalListenerCofnig();
+  setupExternalUpstreamConfig();
+}
+
+// Client sends a clear text HTTP CONNECT message to Envoy.
+// After confirms recevied 200 from proxy, client then sends a
+// clear text HTTP GET message, which will hits the DFP filter in the
+// internal listener and trigger DNS resolving.
+
+TEST_P(HttpsInspectionIntegrationTest, HttpClearInHttpTunnelTest) {
+  setupTestConfig();
+  skip_tag_extraction_rule_check_ = true;
+  initialize();
+
+  // Open clear-text connection.
+  conn_->connect();
+
+  // Sends a HTTP CONNECT message.
+  Buffer::OwnedImpl buffer;
+  std::string connect_msg =
+    "CONNECT www.cnn.com:80 HTTP/1.1\r\n"
+    "Host: www.cnn.com:80\r\n"
+    "\r\n\r\n";
+  payload_reader_->set_data_to_wait_for("HTTP/1.1 200 OK");
+  buffer.add(connect_msg);
+  conn_->write(buffer, false);
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+
+  // Sends a HTTP GET message.
+  auto port_str = std::to_string(fake_upstreams_[0]->localAddress()->ip()->port());
+  std::string get_msg =
+    "GET / HTTP/1.1\r\n"
+    "Host: localhost:"+ port_str  + "\r\n"
+    "\r\n\r\n";
+  buffer.add(get_msg);
+  conn_->write(buffer, false);
+  while (client_write_buffer_->bytesDrained() != connect_msg.length() +
+         get_msg.length()) {
+    dispatcher_->run(Event::Dispatcher::RunType::NonBlock);
+  }
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  ASSERT_TRUE(fake_upstream_connection->waitForData(
+      FakeRawConnection::waitForInexactMatch("GET / HTTP/1.1")));
+  // Sends a response back to client.
+  payload_reader_->set_data_to_wait_for("HTTP/1.1 503 Service Unavailable");
```

Comment:
Why do you get 503? I think you can make it work such that fake upstream responds with 200. It would be good to make sure that routing on internal listener works.